Mathlib/Tactic/ScopedNS.lean

```diff
@@ -1,57 +1,105 @@
-      $[$doc]? $(attr)? scoped $mk:infixl $prec $(n)? $(prio)? $sym => $t)
-  | `($[$doc]? $(attr)? scoped[$ns] $mk:infixr $prec $(n)? $(prio)? $sym => $t) =>
-    `(with_weak_namespace $(mkIdentFrom ns <| rootNamespace ++ ns.getId)
-      $[$doc]? $(attr)? scoped $mk:infixr $prec $(n)? $(prio)? $sym => $t)
-  | `($[$doc]? $(attr)? scoped[$ns] $mk:postfix $prec $(n)? $(prio)? $sym => $t) =>
-    `(with_weak_namespace $(mkIdentFrom ns <| rootNamespace ++ ns.getId)
-      $[$doc]? $(attr)? scoped $mk:postfix $prec $(n)? $(prio)? $sym => $t)
   | `(scoped[$ns] attribute [$[$attr:attr],*] $ids*) =>
+    /- implementation for `Parser.Command.attribute` -/
     `(with_weak_namespace $(mkIdentFrom ns <| rootNamespace ++ ns.getId)
       attribute [$[scoped $attr:attr],*] $ids*)
+  | `($[$docComment]? $(attributes)? scoped%$tk[$ns] $cmd) => do
+    /-
+    implementation for most common commands.
+
+    These commands usually have few optional arguments at the beginning:
+
+    - `optional docComment`: the doc-string
+    - `optional Term.«attributes»`: e.g. `@[simp]`
+    - `Term.attrKind`: is `scoped` or `local`
+
+    We parse the command without these, then add `scoped` manually to control the namespace
+    and add the other optional arguments back.
+    -/
+    let (docCommentPos, attributesPos, attrKindPos) ← match cmd.raw.getKind with
+      | ``Parser.Command.notation
+      | ``Parser.Command.macro_rules
+      | ``Parser.Command.syntax
+      | ``Parser.Command.macro
+      | ``Parser.Command.elab_rules
+      | ``Parser.Command.elab
+      | ``Parser.Command.binderPredicate
+      | ``Parser.Command.mixfix => pure (0, 1, 2)
+      | _ => Macro.throwErrorAt tk "scoped[NS] is not implemented for this command!"
+
+    -- assert the parsed sub-command does not have any of these optional arguments filled
+    -- (e.g. to avoid `scoped[Test] /-- bad -/ local macro ...` from parsing)
+    match cmd.raw[docCommentPos] with
+    | Syntax.node .none `null #[] => pure ()
+    | _ => Macro.throwErrorAt cmd.raw[docCommentPos]
+            s!"unexpected docstring, move it in front of `scoped[NS]`!"
+    match cmd.raw[attributesPos] with
+    | Syntax.node .none `null #[] => pure ()
+    | _ => Macro.throwErrorAt cmd.raw[docCommentPos]
+            s!"unexpected attributes, move them in front of `scoped[NS]`!"
+    let `(Parser.Term.attrKind|) := cmd.raw[attrKindPos]
+      | Macro.throwErrorAt cmd.raw[attrKindPos] "This scoping directive conflicts with scoped[NS]"
+
+    -- parser `optional _` expects a `nullKind` node which
+    -- has no children (case `none`) or one child (case `some _`)
+    let docCommentNode : Syntax := match docComment with
+      | none => Lean.mkNullNode
+      | some doc => Lean.mkNullNode #[doc]
+    let attributesNode : Syntax := match attributes with
+      | none => Lean.mkNullNode
+      | some attr => Lean.mkNullNode #[attr]
+    -- set the `attrKind` to `scoped`
+    let attrKindNode ← `(Parser.Term.attrKind| scoped)
```

Comment:
@digama0 you had `Unhygienic.run` here and I'm not sure if that had any implications I'm not aware of?

Reply:
It's better to use `Unhygienic.run` when the quotation does not have any identifiers or antiquotations inside it because this causes the subterm to be a pure value, so it will be hoisted to a static value.